core/corehttp/gateway_handler.go

```diff
@@ -523,6 +527,25 @@ func (i *gatewayHandler) serveFile(w http.ResponseWriter, req *http.Request, nam
 	http.ServeContent(w, req, name, modtime, content)
 }
 
+func (i *gatewayHandler) serveCBOR(w http.ResponseWriter, r *http.Request, n format.Node) {
+	w.Header().Set("Cache-Control", "public, max-age=29030400, immutable")
+	w.Header().Set("Content-Type", "application/json")
```

Comment:
Not sure about this. We need to make a decision on default behavior before this ships:
-  @warpfork suggested we should aim at returning more user-friendly HTML by default,  similar to explore.ipld.io, for all non-unixfs DAGs. 
   - I kinda agree: we already return custom HTML for directory listings (and not JSON)
   - We want gateways to support everything, not just dag-cbor
- What if for dag-cbor we return JSON only when request had an explicit `Content-Type: application/json` header?
  - This way makes semantic sense (I explicitly ask for JSON representation of dag-cbor) and won't break if we add HTML explorer for IPLD universe as the default response
  - For now, if content-type is missing from request, return error saying that "To preview dag-cbor via gateway, request with `Content-type: application/json` header"

Reply:
@mikeal pointed out that

> there’s a lot of caches out there that aren’t smart about this and you’ll get responses from cache that don’t match the accept header

This is really unfortunate, but likely relevant concern.

I was thinking maybe gateway code could consider file extension if there is one and infer Content-Type from there ? E.g. if content targeted is `ipfs://${cid}/metadata.json` in dag-cbor format it's probably reasonable to assume that content type `application/json` should be used. In fact that seems like a reasonable approach to dag-pb as well.

Although I realize this is getting uncomfortable.

Reply:
I'd be _ok_ with streaming the bytes as `application/cbor` and add a querystring parameter `?json` or `?encoding=json` to have it converted to JSON with a `application/json` header.

IMHO we should be serving data from the gateway that is by default the easiest format to consume from a web application. JSON is that. Also, right now we have NO support, and this is significantly better than nothing and what's here _is_ consistent with the CLI.

So I'd prefer to keep what's happening here and maybe add `?encoding=raw` for getting the raw cbor bytes.

I like @gozala's idea but it feels like too many special cases.

Reply:
> you’ll get responses from cache that don’t match the accept header

Reply:
We really should keep complexity at bay and avoid special-casing different DAG types too much.
I propose we  introduce DAG CBOR support in (IMO) the least controversial way:

- Stream `application/cbor` by default
  - Returning the original, space-efficient format makes more sense than JSON
    - Rationale: if someone wants JSON by default, then they should use `dag-json`
  - IIUC [dag-cbor](https://github.com/ipld/specs/blob/master/block-layer/codecs/dag-cbor.md) is a subset of cbor, so this is a valid content-type (I'd like to avoid introducing proprietary one if we don't have to). 
-  JSON requires conversion step, so we should support  `application/json` as an explicit opt-in:
   - requesting `Content-Type: application/json`
     - Rationale: standards matter and we can't be sloppy if we want to be respected by serious vendors and standards bodies
   - passing `?enc=json` parameter (as a fallback way for use in poorly written software)
     - Rationale:  the same parameter could be re-used for returning JSON directory listings in https://github.com/ipfs/go-ipfs/issues/7552, keeping our conventions neat.

Reply:
Just require the header and the querystring. The problem you’ll run into with it only being a fallback is that the user, client developer, and the meddling proxy/cache are different actors and can’t always coordinate whether or not to explicitly opt into a feature.

Reply:
Good point on the address bar. It would actually have inconsistent behavior based on your browser plugins because everyone with JSON viewer plugin includes the content type in browser requests.

Reply:
Let's continue in https://github.com/ipfs/in-web-browsers/issues/182